torch/distributed/_tensor/ops/tensor_ops.py

```diff
@@ -465,77 +508,102 @@ def cat_rule(op_schema: OpSchema) -> OutputSharding:
+            is_tensor_dim_sharded(spec, dim=dim) or is_tensor_partial(spec)
+        ):
             need_reshard = True
             tensor_list_specs_after.append(
                 DTensorSpec(
                     mesh=spec.mesh,
-                    placements=unshard_tensor_dim(spec.placements, dim=dim),
+                    placements=replicate_tensor_dim(spec.placements, dim=dim),
                     tensor_meta=spec.tensor_meta,
                 )
             )
         else:
             tensor_list_specs_after.append(spec)
+
     tensor_list_specs = tensor_list_specs_after
 
-    # TODO: currently einop rule requires every character
-    # in result notation must have appeared in inputs
-    # so we temporarily design cat notation as
-    # "aij,bij->aij". Once we modify this requirement,
-    # we can switch to the more logically reasonable notation
-    # "aij,bij->cij"
-    alphabet = "abcdefghijklmnopqrstuvwxyz"
-    einop_notation_list = []
-
-    l = len(tensor_list_specs)
-    free_dim = alphabet[l : l + ndim - 1]
-    for i, spec in enumerate(tensor_list_specs):
-        if spec.ndim == ndim:
-            # rewrite concat dim
-            dim_word = free_dim[:dim] + alphabet[i] + free_dim[dim:]
-            einop_notation_list.append(dim_word)
+    # align non-cat dimensions placements based on reshard cost
+    non_empty_specs = [spec for spec in tensor_list_specs if not is_empty(spec)]
+    mesh = non_empty_specs[0].mesh
+    ndim = non_empty_specs[0].ndim
+    new_placements: List[Placement] = []
+    for mesh_dim in range(mesh.ndim):
+        # compute the minimum cost of resharding on this mesh_dim
+        if any(
+            spec.placements[mesh_dim] != non_empty_specs[0].placements[mesh_dim]
+            for spec in non_empty_specs
+        ):
+            # only reshard if there is a mismatch
+            need_reshard = True
+            reshard_cost = []
+            for shard_dim in range(ndim):
+                # compute the cost of resharding on this shard_dim
+                cost: float = 0.0
+                for spec in non_empty_specs:
+                    global_shape = spec.shape
+                    if global_shape[shard_dim] < mesh.size(mesh_dim):
+                        # found one tensor where the shard_dim is smaller than
+                        # mesh_dim. In this case, we cannot shard on this shard_dim,
+                        # and hence set cost to infinity.
+                        cost = +float("inf")
+                    elif (
+                        is_tensor_dim_sharded(spec, dim=shard_dim)
+                        or prod(global_shape) == 0
+                    ):
+                        continue
+                    else:
+                        local_shape = compute_local_shape(
+                            global_shape, spec.mesh, spec.placements
+                        )
+                        cost += prod(local_shape) * spec.mesh.size(mesh_dim)
```

Comment:
is this an `allgather` cost I assume? what if the current spec is `replicate` where it does not need allgather?

Reply:
Oh I thought this is reshard cost copied from here. Let me know this is wrong. 

https://github.com/pytorch/pytorch/blob/4582ceb2c4a7316f09afd471a97d910347d21f01/torch/distributed/_tensor/ops/common_rules.py#L181-L190

Reply:
I see, I think it's fine to use this for now, we probably need to standardize this logic later.